xformers/components/attention/csrc/cuda/attention.cu

```diff
@@ -174,6 +208,55 @@ __device__ void compute_final_mult(
   }
 }
 
+template <typename scalar_t, typename vec_t, int kBlockSizeK, int kBlockSizeQ>
+//__device__ __forceinline__ void apply_masking(
+__device__ void apply_masking(
+    scalar_t s_delta[kBlockSizeQ][kBlockSizeK],
+    at::PhiloxCudaState philox_args,
+    int64_t global_offset,
+    int64_t N,
+    scalar_t p,
+    int64_t col_offset) {
+  // strategy: initialize the rng so that each element in the attention
+  // matrix has its own subsequence, so that we can easily retrieve
+  // the element during backward
+
+  curandStatePhilox4_32_10_t state;
+  auto seeds = at::cuda::philox::unpack(philox_args);
+
+  // we will always sample 4 random floats at a time
+  // as it's more efficient
+  constexpr int kSampled = 4;
+
+  // because the forward and the backward have different
+  // access patterns, we round the rng offset so that it's
+  // a multiple of kSampled, and add the delta needed
+  int delta = col_offset & (kSampled - 1);
+  col_offset = col_offset - delta;
+
+#pragma unroll
+  for (int64_t q_item_idx = 0; q_item_idx < kBlockSizeQ; q_item_idx++) {
+#pragma unroll
+    for (int64_t k_item_idx = 0; k_item_idx < kBlockSizeK;
+         k_item_idx += kSampled) {
+      int64_t offset = global_offset + q_item_idx * N + k_item_idx + col_offset;
+      // ideally we would use the code below, but initializing the rng
+      // takes a significant portion of time. So we instead modify the seed
+      // so that each thread has a different seed. This has fewer statistical
+      // guarantees than by doing it properly, but is much faster
+      curand_init(
+          std::get<0>(seeds), offset, std::get<1>(seeds) + delta, &state);
+      // curand_init(std::get<0>(seeds) + (offset << 8) + std::get<1>(seeds), 0,
```

Comment:
old API I suppose, still needed around ?

Reply:
that should be removed actually, was a workaround to get faster generation back when it was slower (but had fewer guarantees regarding randomness).